src/Interpreter/Mulang.hs

```diff
@@ -0,0 +1,387 @@
+  returnValue <- callCC $ \(returnCallback) -> do
+    put (context { scopes = functionEnv, currentReturnCallback = returnCallback })
+    evalExpr body
+    return nullRef
+  modify' (\c -> c { scopes = scopes context
+                   , currentReturnCallback = currentReturnCallback context
+                   , currentRaiseCallback = currentRaiseCallback context
+                   , currentException = currentException context
+                   })
+  return returnValue
+
+findFrameForName :: String -> Executable Reference
+findFrameForName name = do
+  maybe (raiseString $ "Reference not found for name '" ++ name ++ "'") return
+    =<< findFrameForName' name
+
+findFrameForName' :: String -> Executable (Maybe Reference)
+findFrameForName' name = do
+  framesRefs <- gets scopes
+  frames :: [(Reference, Map String Reference)] <- forM framesRefs $ \ref -> do
+    dereference ref >>= \value -> case value of
+      (MuObject context) -> return (ref, context)
+      v -> error $ "Finding '" ++ name ++ "' the frame I got a non object " ++ show v
+
+  return $ fmap fst . find (Map.member name . snd) $ frames
+
+
+findReferenceForName :: String -> Executable Reference
+findReferenceForName name = do
+  ref <- findFrameForName name
+  (MuObject context) <- dereference ref
+  return $ context Map.! name
+
+getMaxKey :: Map k a -> Maybe k
+getMaxKey m = case Map.maxViewWithKey m of
+  Just ((k, _a), _) -> Just k
+  _ -> Nothing
+
+nullRef = Reference 0
+
+dereference' :: ObjectSpace -> Reference -> Value
+dereference' _ (Reference 0) = MuNull
+dereference' objectSpace ref = do
+  fromMaybe (error $ "Failed to find ref " ++ show ref ++ " in " ++ show objectSpace) .
+    Map.lookup ref $
+    objectSpace
+
+dereference :: Reference -> Executable Value
+dereference ref = do
+  objectSpace <- gets globalObjects
+  return $ dereference' objectSpace ref
+
+updateGlobalObjects f context =
+  context { globalObjects = f $ globalObjects context
+          }
+
+updateLocalVariables f context =
+  context { scopes = f $ scopes context
+          }
+
+incrementRef (Reference n) = Reference $ n + 1
+
+createReference :: Value -> Executable Reference
+createReference value = do
+  nextReferenceId :: Reference <- gets (fromMaybe (Reference 10) . fmap incrementRef . getMaxKey . globalObjects)
+  modify (updateGlobalObjects $ Map.insert nextReferenceId value)
+  return nextReferenceId
+
+currentFrame :: Executable Reference
+currentFrame = do
+  gets (head . scopes)
+
+setLocalVariable :: String -> Reference -> Executable ()
+setLocalVariable name ref = do
+  frame <- currentFrame
+  updateRef frame (addAttrToObject name ref)
+
+addAttrToObject :: String -> Reference -> Value -> Value
+addAttrToObject k r (MuObject map) = MuObject $ Map.insert k r map
+addAttrToObject k _r v = error $ "Tried adding " ++ k ++ " to a non object: " ++ show v
+
+putRef :: Reference -> Value -> Executable ()
+putRef ref val = do
+  modify $ updateGlobalObjects $ Map.insert ref val
```

Comment:
I'd prefer to avoid overuse of application operation, and favour composition operation instead. E.g.:

```haskell
modify . updateGlobalObjects $ Map.insert ref val
modify . updateGlobalObjects . Map.insert ref $ val
```


My rule of thumb is one `$` per line.

Reply:
Also, Mulang code tends to favor point-free style and avoid unnecessary `do`-blocks: 

```
putRef ref  = modify . updateGlobalObjects . Map.insert ref
```